File: practicas/Semana-06/Agua/Agua.java
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.Random;

//A shared resource/class
class Shared {
    AtomicInteger cH;
    AtomicInteger cO;
    Semaphore SemH;
    Semaphore SemO;

    public Shared() {
       cH = new AtomicInteger();	// Inits variable in zero
       cO = new AtomicInteger();
       // creating a Semaphore object with number of permits 0
       SemH = new Semaphore(0);
       SemO = new Semaphore(0);
    }

    public int getH() {
       return cH.get();
    }

    public int getO() {
       return cO.get();
    }

    public void IncH() {
       cH.getAndIncrement();
    }

    public void DecH() {
       cH.getAndDecrement();
    }

    public void IncO() {
       cO.incrementAndGet();
    }

    public void DecO() {
       cO.decrementAndGet();
    }

    public void SignalH() {
       this.SemH.release();
    }

    public void SignalO() {
       this.SemO.release();
    }

    public void WaitH() {
       try {
          this.SemH.acquire();
       } catch ( InterruptedException e ) {
       }
    }

    public void WaitO() {
       try {
          this.SemO.acquire();
       } catch ( InterruptedException e ) {
       }
    }

}
  
class Particle extends Thread {
    Random r;
    int number;
    Shared resources;

    public Particle( int i, Shared res ) {
       this.number = i;
       this.r = new Random();
       this.resources = res;
    } 
  
    @Override
    public void run() {
       if ( r.nextBoolean() ) {		// Create a new particle, randomly select hydrogen and oxygen
          System.out.println("Starting O thread " + this.number + " ... cO(" + resources.getO() + "), cH(" + resources.getH() + ")" );
          this.O();
       } else {
          System.out.println("Starting H thread " + this.number + " ... cO(" + resources.getO() + "), cH(" + resources.getH() + ")" );
          this.H();
       }

    }

    public void O() {
       if ( resources.getH() > 1 ) {
          System.out.println("\t O particle making water :) ... cO(" + resources.getO() + "), cH(" + resources.getH() + ")" );
          resources.DecH(); resources.DecH();
          resources.SignalH();
          resources.SignalH();
       } else {
          resources.IncO();
          resources.WaitO();
       }
    }

    public void H() {
       if ( (resources.getH() > 0) && (resources.getO() > 0) ) {
          System.out.println("\t H particle making water :) ... cO(" + resources.getO() + "), cH(" + resources.getH() + ")" );
          resources.DecH();
          resources.DecO();
          resources.SignalH();
          resources.SignalO();
       } else {
          resources.IncH();
          resources.WaitH();
       }
    }

} 
  
// Driver class 
public class Agua {
    public static void main(String args[]) throws InterruptedException {
        Shared agua = new Shared();

        Thread particles[] = new Thread[ 100 ];

        for ( int i = 0; i < 100; i++ ) {
           particles[ i ]  = new Particle( i, agua );
           particles[ i ].start();
        }

        // waiting for threads
        for ( int i = 0; i < 100; i++ ) {
           try {
              particles[ i ].join();
           } catch ( InterruptedException e) {
           }
 
        }
    }

}

File: practicas/Semana-06/Agua/Agua.py
import threading
import sys
import time
import random

"""
   Shared class contains resources to resolv H2O problem
   It has
      - two integers to count hydrogen and oxygen particles
      - two locks to make all integers operations atomic
         - return, increment, and decrement are executed atomically using these locks
      - two semaphores to stop particles when it is not possible to make water
"""
class Shared:

	def __init__( self ):	# Constructor
		self.cH = 0
		self.cO = 0
		self.SemH = threading.Semaphore( 0 )
		self.SemO = threading.Semaphore( 0 )
		self.lockH = threading.Lock()
		self.lockO = threading.Lock()

	def getH( self ):
		with self.lockH:	# Only one thread can pass this lock
			return self.cH

	def getO( self ):
		with self.lockO:
			return self.cO

	def decH( self ):
		with self.lockH:
			self.cH -= 1

	def decO( self ):
		with self.lockO:
			self.cO -= 1

	def incH( self ):
		with self.lockH:
			self.cH += 1

	def incO( self ):
		with self.lockO:
			self.cO += 1

	def signalH( self ):
		self.SemH.release()

	def signalO( self ):
		self.SemO.release()

	def waitH( self ):
		self.SemH.acquire()

	def waitO( self ):
		self.SemO.acquire()


"""
   Agua is where all particles are created using a thread for each one
   It is an extension for the threading.Thread class and we need to overcharge constructor and run methods
"""
class Agua ( threading.Thread ):

	def __init__( self, threadID, resources ):
		threading.Thread.__init__( self )
		self.threadID = threadID
		self.resources = resources

	def run(self):	# Create randomly hydrogens and oxygens
		r = random.randint( 1, 10 )
		if ( 0 == (r % 2) ):
			print( "Starting O thread " + str( self.threadID ) + " ... cO(" + str(self.resources.getO()) + "), cH(" + str(self.resources.getH()) + ")" )
			self.O()
		else:
			print( "Starting O thread " + str( self.threadID ) + " ... cO(" + str(self.resources.getO()) + "), cH(" + str(self.resources.getH()) + ")" )
			self.H()

	def H( self ):	# Represent an hydrogen particle
		if ( (self.resources.getO() > 0) and (self.resources.getH() > 0) ):
			print( "\tH particle making water :) ... " + str( self.threadID ) + " ... cO(" + str(self.resources.getO()) + "), cH(" + str(self.resources.getH()) + ")" )
			self.resources.decH()
			self.resources.decO()
			self.resources.signalO()
			self.resources.signalH()
		else:
			self.resources.incH()
			self.resources.waitH()

	def O( self ):	# Represent an oxygen particle
		if ( self.resources.getH() > 1 ):
			print(  "\tO particle making water :) ... " + str( self.threadID ) + " ... cO(" + str(self.resources.getO()) + "), cH(" + str(self.resources.getH()) + ")" )
			self.resources.decH()
			self.resources.decH()
			self.resources.signalH()
			self.resources.signalH()
		else:
			self.resources.incO()
			self.resources.waitO()

def main():
	threads = []
	resources = Shared();

	for i in range( 100 ):
		t = Agua( i, resources )
		threads.append( t )
		t.start()

	for t in threads:
		t.join()

if __name__ == "__main__":
    main()

